### ISSUE TITLE
Lack of `install_requires`

### ISSUE THREAD

**Issue by dopplershift** [sphinx-gallery/sphinx-gallery#192]
I know `setup.py` notes that not having `install_requires` is a deliberate decision, but is it possible to revisit that decision?

Right now `pip install sphinx-gallery` results in an incomplete install--I still need to manually install sphinx, matplotlib, and pillow to have a functional install. While sphinx and matplotlib are a given for my project, I need to now add `pillow` as a manual install step to my documentation CI builds--and it will break any time dependencies update in the future. Not to mention this becomes a much bigger headache should there be any version-specific needs in the future.

**Comment by dopplershift** [comment]
Documenting this fact prominently would be a nice start.

Putting things in my own `requirements.txt` doesn't solve my core issue here: that I *manually* must do the job of a package manager and stay up to date on what dependencies must be installed to have sphinx-gallery function properly. I've never before encountered a Python package that does this and doesn't put things in `install_requires`. (`install_requires` is considered best practice here: https://packaging.python.org/distributing/#install-requires)

It feels like wrong software engineering practice to me, since not only do I have to repeat information that exists and is used elsewhere, but I'm also codifying in the configuration for my own project information that is completely unrelated.

I don't know the problems that led to this solution, so at the very least I'd suggest better commenting/documenting the actual problem rather than saying "install_requires breaks stuff" (the pull request that made this change literally says nothing). Ideally, I'd revisit the problem and find a solution that doesn't require doing something so far from normal in the Python ecosystem--but obviously that's not my call.

**Comment by dopplershift** [comment]
To be clear, I have no problem using a `requirements.txt`. My problem is that mine would have to be something like:
```
numpy
scipy
matplotlib
sphinx
sphinx-gallery
# The following are needed to make sphinx-gallery work
pillow
 # outdated
joblib
```
So now I have to maintain `requirements.txt` not just for the things I want in the environment, but I have to make sure I'm up-to-date on what changes you guys make, like no-longer needing joblib,  or adding new dependencies. It feels really wrong to be putting in that file things that I don't directly want, but are unlisted requirements, from pip's perspective, for sphinx-gallery.

**Comment by lesteve** [comment]
For the record, I looked at the doc and it turns out that you can include sphinx_gallery requirements in your own requirements file like this:

```
# your requirements go here
numpy
scipy

# sphinx-gallery requirements
-r https://raw.githubusercontent.com/sphinx-gallery/sphinx-gallery/master/requirements.txt
```

**Comment by lesteve** [comment]
To be fair we are probably on the conservative side with regards to adding dependencies in sphinx-gallery, so I do not expect them to change very often.

Also pip has recently improved but until very recently (and the manylinux wheels) it used to be a pain for compiled extensions (it is good bearing in mind that sphinx-gallery originates from scikit-learn people). Look at https://github.com/scikit-learn/scikit-learn/issues/6665 for a recent bug where mixing pip and conda for installing numpy yields to a very peculiar behaviour.

**Comment by dopplershift** [comment]
Thanks for the tip--I'll keep that in mind.

I just thought I'd make the case for using what IMO is more standard; it's not then end of the world to put pillow in the list of things I need in CI (I already need Cython there as well).

Re: conda mixed with pip--that's its own special version of hell.

**Comment by anntzer** [comment]
If you do not want pip to ever mess your dependencies, you should just set the PIP_NO_DEPS environment variable, or similarly set `~/.config/pip/pip.conf` appropriately (https://pip.pypa.io/en/stable/user_guide/#configuration).  This way, you will have no problem whenever installing a package, regardless of whether it declares its requirements correctly or not.

I would thus still suggest that sphinx-gallery does declare install_requires properly, so that those who do want pip to install dependencies can benefit from it.

**Comment by QuLogic** [comment]
> we deliberately disabled the `install_requires` as it broke your developer environment for mayavi.

Isn't mayavi optional? Why not declare it as an [extra dependency](https://setuptools.readthedocs.io/en/latest/setuptools.html#declaring-extras-optional-features-with-their-own-dependencies), not a hard requirement?

> Getting mayavi to work with sphinx-gallery and other packages is a pain, our travis CI file shows the extra steps required to get it to work.

All I see are peculiarities of the conda packages there, not necessarily pip. (I don't doubt that there are though, considering it's based on VTK+.)

**Comment by Titan-C** [comment]
> Isn't mayavi optional?

Yes is optional. But is you have mayavi installed you need some
specific versions in its dependencies for it to work. We don't want to
break that specific config when installing sphinx-gallery.

>> Getting mayavi to work with sphinx-gallery and other packages is a pain, our travis CI file shows the extra steps required to get it to work.
>
> All I see are peculiarities of the conda packages there, not necessarily pip. (I don't doubt that there are though, considering it's based on VTK+.)

It is true that the situation is getting better as time goes on, and
control of our virtual environments is simpler nowadays. As such our
travis file has gotten simpler. It is nevertheless important to keep in
mind the ordering. You install first sphinx-gallery, then mayavi. As it
has to downgrade some packages, and then(not true anymore) you have to
reupgrade other dependencies(pyface).

I think, we could fix our requirements.txt to the minimum supported
versions of our dependencies in the hope pip never updates dependencies
when installing the gallery.

**Comment by QuLogic** [comment]
Pip doesn't update dependencies unless you ask it to, or maybe if there's a conflict (never had to worry about that before, so not totally sure). I don't see anything in the documented list of Mayavi2 requirements that would conflict with the list in `requirements.txt` (or their requirements), except for ensuring that Matplotlib and Mayavi2 use the same backends.

> As such our travis file has gotten simpler.

The Travis file was just as simple when this issue was opened, though.

**Comment by anntzer** [comment]
Not sure what you are worried about when it comes to installation order: anything seems to work for me:

```
# Installing vtk without conda would be a bit painful.
conda create -n testenv python=2.7
source activate testenv
# Use recent versions of packaging tools.
pip install -U setuptools pip

### The following two installations can be done in either order.

conda install -c conda-forge mayavi 'pyqt<5'

# from the sphinx-gallery source, with the following patch applied (declaring dependencies):
# diff --git a/setup.py b/setup.py
# index 1341975..e28c556 100644
# --- a/setup.py
# +++ b/setup.py
# @@ -15,7 +15,7 @@ with codecs.open('README.rst', mode='r', encoding='utf-8') as f:
#  
#  # XXX : don't force requirements in setup.py as it tends to break
#  #       people's setups
# -install_requires = []
# +install_requires = ['sphinx', 'matplotlib', 'pillow']
#  
#  if os.environ.get('READTHEDOCS', None) == 'True':
#      # So that we can build documentation using seaborn

pip install .

###

mayavi2  # Works regardless of install order.
```

**Comment by larsoner** [comment]
Nowadays `pip` allows easy installation of everything, and I tend to find more harm than good done by omitting things from the requires list. I'd be fine adding `matplotlib`, `sphinx`, and `pillow` as dependencies (and adding a CI entry that things can actually work with this setup). Mayavi should be excluded, or added as an optional so people can do `pip install sphinx_gallery[mayavi]`.

**Comment by GaelVaroquaux** [comment]
> I'd be fine adding matplotlib, sphinx, and pillow as dependencies

+1

> (and adding a CI entry that things can actually work with this setup).

+1

> Mayavi should be excluded, or added as an optional so people can do pip install sphinx_gallery[mayavi].

+1

**Comment by choldgraf** [comment]
note that this is now largely being addressed in @larsoner 's PR: #445
